### DISCUSSION TITLE
Compile only UI components used

### DISCUSSION THREAD

**Discussion by mnzsss** [vercel/turborepo#4696] [Help] [answered]
Hello I have follow the basics design system example and get a problem.

I have the `web` project created with Next.js and migrate a existing component library to `packages` in Turbo, so I can't compile only components if I use in one page. Basicly the component library structure is:

```bash
├── apps/
│   └── web/
│       ├── pages
│       └── ...
└── packages/
    └── ui/
        ├── components/
        │   └── index.ts
        └── ...
```

The problem is the library ui can't compile when I run `next dev` with this error:

![image](https://user-images.githubusercontent.com/51327920/234364026-1172103e-e1ed-4e42-bfe6-570ba05b39f9.png)

I use in my `tsconfig.json`:

```json
// packages/ui/tsconfig.json
{
  "extends": "@beyounglabs/ts-config/react.json",
  "include": ["."],
  "exclude": ["dist", "build", "node_modules"],
  "compilerOptions": {
    "baseUrl": "./src",
    "paths": {
      "ui/*": ["./src/*"]
    }
  }
}
```

```json
// packages/ui/tsconfig.json
{
  "extends": "@beyounglabs/ts-config/nextjs.json",
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"],
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "*": ["*"],
      "ui/*": ["../../packages/ui/src/*"]
    }
  }
}

```

If I use the the `transpilePackages` from Next.js in `next.config.js` the package compiled but compile all components and others files if I'm not using and increase a big bundle of my app

![Screenshot from 2023-04-25 12-42-34](https://user-images.githubusercontent.com/51327920/234365554-d5f427d5-9e59-4c5b-959c-777217ea6c1d.png)

Has any way to compile only components used in `web` whithout `transpilePackages`?

**Comment by anthonyshew** [comment, marked as answer]
This is likely because you're using a barrel file export.

Check out the new `npx create-turbo@latest` where we export each component from `ui` as its own entrypoint. This forces stronger module boundaries so you (or your bundler) can't bloat your bundles.
